### ISSUE TITLE
Errant Late Charge Fix

### ISSUE THREAD

**Issue by kdjstudios** [kdjstudios/SABillingGitlab#1158]
In GitLab by @tim.traylor on Oct 3, 2018, 07:08

This ticket is to address both of the late charge fixes for hosted and Answernet (issues #1146 and #1151).

1. Generate a credit to offset the late charge.
1. We will provide Aavaz with a list of sites and the staged fee that should be used to credit the charge
1. Aavaz will generate the staged fees in the amount listed in the spreadsheets provided to us
1. Aavaz will create a new collection (titled bad_late_charges) which consists of the data from the spreadsheets as well a new column titled 'cycle_processed_date'.

The invoice generation code will temporarily be updated to perform the following:
1. Every account goes through the 'check for late charges' process.
1. The new code will lookup the account in the  bad_late_charges collection and determine it is an affected account.
1. If it is an affected account:
    1. If the  'cycle_processed_date' < the current billing cycle date then we proceed with normal late charging processing.
    1. Otherwise, there is an unpaid balance  <=  the amount of the bad late charge then the account will NOT be charged late charges for the current invoice
    1. If there is an unpaid balance > the amount of the bad late charge then proceed with normal late charging process.
    1. Write the billing cycle date to the cycle_processed_date for this account.
1.  If it not an affected account then we proceed with normal late charging processing.

[answernet_bad_late_charges.csv](/uploads/577352728d58aefe042b3b0ab17e2d91/answernet_bad_late_charges.csv)

[hosted_bad_late_charges.csv](/uploads/604583b0c1c0d9f14d693cf68eb48f28/hosted_bad_late_charges.csv)

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Oct 3, 2018, 07:11

changed the description

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Oct 3, 2018, 07:12

changed the description

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Oct 3, 2018, 07:12

changed the description

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Oct 3, 2018, 07:13

changed the description

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Oct 3, 2018, 07:14

changed the description

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 3, 2018, 08:39

@smasih @pchaudhary @amishra @tim.traylor 

I am sending out 4 external emails and 18 internal emails for each site accordingly. If I do not receive a response before the end of day today I will follow up on each email thread accordingly. 

Team, please be prepared for 22 new tickets, one for each site/client.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 3, 2018, 11:15

Emails have been sent to both internal and external. I have created tickets 1161 to 1182 for each client/site accordingly.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Oct 4, 2018, 02:49

#### Consolidated list of ticket and their status.

----

| **Ticket** | **Response Status**                                           | **Progress Status** |
|------------|---------------------------------------------------------------|---------------------|
| #1162      | Deb APPROVED and CODE is new Staged Fee # 564                                           |                     |
| #1163      | No Response til now                                           |                     |
| #1164      | Jo Ann APPROVED and CODE is 1031                                           |                     |
| #1165      | Gaylan Approved but there are [accounts he already credited the fees](http://gitlab.aavaz.biz/AnswerNet/SABilling/uploads/5f409c6eb6a425152385a4d9153ae72d/Late+Fee+Credits+for+Keener.xlsx)                                           |                     |
| #1166      | Michelle already set up a staged fee to correct this account. |                     |
| #1167      | Lakeisha approved but waiting for credit staged fee code      |                     |
| #1168      | Amecia APPROVED and CODE is 7782                              |                     |
| #1169      | Tobey APPROVED and CODE is 7980                                           |                     |
| #1170      | Valerie APPROVED and CODE is 9904                             |  :ballot_box_with_check:                   |
| #1171      | Kimberly APPROVED and CODE is 7882                            | :ballot_box_with_check:                     |
| #1172      | Jesus APPROVED and CODE is 7980                                           |                     |
| #1173      | No Response til now                                           |                     |
| #1174      | Laura APPROVED and CODE is `044 Late Chg Adj 7882`                                            |                     |
| #1175      | Tobey APPROVED and CODE is 7882                                           |                     |
| #1176      | Jerri Herrguth APPROVED and CODE is 7882                      | :ballot_box_with_check:                    |
| #1177      | Sarah APPROVED and CODE is `112 Late Fee Credit 7882`                                           |                     |
| #1178      | Lettice APPROVED and CODE is 7980                                          |                     |
| #1179      | Melissa  APPROVED and CODE is 7882                            |  :ballot_box_with_check:                   |
| #1180      | Elizabeth  APPROVED and CODE is 990*(wait)                           |                     |
| #1181      | Denise will provide an update soon                                           |                     |
| #1182      | Tobey APPROVED and CODE is 304                                           |                     |
| #1185      | Jerri APPROVED and CODE is 7882                                           |                     |

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Oct 4, 2018, 08:40

updated #1179

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Oct 4, 2018, 08:51

added 5h of time spent at 2018-10-04

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 4, 2018, 10:53

I have updated all tickets and requested update on tickets that we have not received a response on. Please note I missed created a gitlab ticket for Santa Rosa so that has also been completed #1185. I will continue to update each individual ticket as we continue to receive updates.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Oct 5, 2018, 01:25

Updated #1171

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Oct 5, 2018, 01:34

Updated #1170

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Oct 5, 2018, 02:23

Updated #1168

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Oct 5, 2018, 02:31

Updated #1176

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Oct 5, 2018, 04:04

Updated #1169

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Oct 5, 2018, 04:24

Update #1172

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Oct 5, 2018, 04:42

Updated #1175

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Oct 5, 2018, 04:50

Updated #1177

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Oct 5, 2018, 05:13

Updated #1182

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Oct 5, 2018, 05:18

Updated #1185

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Oct 5, 2018, 06:40

Updated #1164

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Oct 5, 2018, 07:03

Updated #1162

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Oct 5, 2018, 07:16

Updated #1167

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 5, 2018, 07:54

@smasih @pchaudhary @tim.traylor @amishra 

Thanks everyone for compiling the list of accounts with invalid late fees. However, we are running into a concern that possibly a majority of these accounts are valid late fees. May you please explain how you determined if an account had an valid vs invalid late fee assess on it?

NOTE: Please reference the accounts in #1181 and #1176 as both of these sites are reporting 75% or more accounts were valid.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 5, 2018, 10:01

@smasih wrote:

> - we took all the billing cycles of the site b/w `2018/09/01 -to- 2018/10/02`
> 
> - mostly there was only one billing cycle
> 
> - and from the billing cycle we extracted all the invoices linked to it
> 
> - now account's `net terms` comes into play
> 
> - we traversed each invoice and extracted the amount of line item labelled as "late Charge"
> 
> - now the next step was to check if the fee was valid or not
> 
>     - if net terms of the account is `0` then system directly charges a late fee on the remaining outstanding
> 
>     - if net terms are not `0` then for eg: 15
> 
>     - system checks if previous invoices were paid before "current Invoice date - 15 days"
> 
>     - so we used the net terms and calculated which invoices were not fully paid before the "current Invoice date - net terms"

However after checking a random group of accounts for Orlando, Toronto and Sarasota; it appears that possibly the Net Terms = 0 check to generate the list did not working correctly. I have listed four of the accounts below that are all setup to with NET=0 and were charged a valid late fee due to the accounts having a previous balance. However, all the accounts ended up on the list. 

118-SR5155 - Previous Balance $114.00 - [invoice_118-75522.pdf](/uploads/44df8b8b4aa8e83c9d2e014d0d631cbe/invoice_118-75522.pdf)

118-SR6553 - Previous Balance $140.50 - [invoice_118-75751.pdf](/uploads/3528199b5427278f1cdaa6012efe6e33/invoice_118-75751.pdf)

062-SER1091 - Previous Balance 1194.10 - [invoice_062-40325.pdf](/uploads/1deba08ab4ec685ea23de48db824a0eb/invoice_062-40325.pdf)

062-SER1010 - Previous Balance 6942.46 - [invoice_062-40112.pdf](/uploads/9a98c27e58ac7d1de1b25638bd6f7c89/invoice_062-40112.pdf)

045-S06531 Previous Balance 378.05 - [invoice_045-81576.pdf](/uploads/78f74f2b9e500d0c08cebcb0190f8867/invoice_045-81576.pdf)


I also checked accounts at El Paso that had NET=15 to confirm that those were showing correctly too. However, it would appear those late fees were valid too? So I am at a loss on how we determined the list of accounts, because it does not appear to match what was stated above. 

025-M6558 - Previous Invoice date 9/1/18; Payment date 9/19/18 - 4 days late. - [invoice_025-52805.pdf](/uploads/e4382591e48aa6a8bd6dc2ab54a498d7/invoice_025-52805.pdf)

025-E5671 - Previous Invoice date 9/1/18; Payment date 9/18/18 - 3 days late. -[invoice_025-52766.pdf](/uploads/813c35899526be9db4e9dcf61d68ef5e/invoice_025-52766.pdf)


Sadly I don't have more time to keep spot checking accounts today. Above are just the accounts that I appear are valid late fees. Please note that there was probably just as many accounts I checked that were on the list and had invalid late fees, so I am not saying everything is wrong; but from what I can see so far is the following:

1. Our lists of invalid late fees has quite a few of valid late fees. We will want to discuss this with @tim.traylor on Monday on what to do.

2. Operations at some sites possibly did not check the list before approving and we may now have accounts with invalid late fee credits on them.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 5, 2018, 10:15

@smasih @pchaudhary @amishra @tim.traylor 

Marked for discussion. Please review this with @tim.traylor on Monday Morning.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Oct 9, 2018, 01:11

Updated #1180

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Oct 9, 2018, 01:11

added 30m of time spent at 2018-10-09

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 9, 2018, 09:59

Discussed today. The Aavaz team is going to look further into this and find what was causing valid accounts to show as invalid. We will also need an updated list of accounts with invalid late fees.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Oct 9, 2018, 13:26

@kdjstudios @tim.traylor 

Here is a full list or accounts to verify. Hopefully this will provide full visibility into what needs to be done. 

The description sheet - gives an overview of each of the sheets. 

We are waiting on the definition of Net 0. But we can look at the code and treat it as it is being handled there. 

[account_with_late_feesanswernet23__copy_.xlsx](/uploads/45b7f6c23697a2cfffe93f1aeaa72dfc/account_with_late_feesanswernet23__copy_.xlsx)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 9, 2018, 14:00

@smasih 

Thank you for the update. Please look into how SAB is currently processing Net 0. @tim.traylor has been in meetings all day and I have not been able to discuss this with him. I think we will need to have this discussed further in-depth tomorrow morning. 

Just to notate from what my personally understanding of what Net 0 meant is that the payment is due 'immediately'; however a late charge would not be assessed unless the next invoice was generated and only if there was a previous/remaining balance. This way it works with any sort of billing period/cycle.

However, after researching online, it appears the 'Net 0' is really left open for interpretation. Companies commonly use it when a client is on a form of electric auto-pay and the payment is processed immediately and automatically without client interaction. This way a late fee is only added if that payment was declined.

Most online research I have gathered is that 'Net 0' is really a 'Net 30'; but since it says "Due upon receipt" on the invoice it helps motivate people to pay quicker. (NOTE: From what I could infer that also they are on a billing cycle of 30 days; Thus how I got my understanding above.)

Next most popular is that it means 48 hours; which would really be a 'Net 2'. It would also appear that generally there is additional text on the invoice that states this too. 

While other companies use 1 week, 2 weeks, 4 weeks, etc. Just about any period. Some in addition to that have 'grace periods' in combination to these, before a late fee is charged.

- Just an additional note; overall it seems accounts that are not receiving their invoices immediately(Email) should not be setup for 'net 0'. This is because physical mailing has delays.

- One personal thought too, is that extenuating circumstances. As much as I don't want to admit it. But I do believe that commonly invoices are finalized and sent out after the invoice date; which in turn can cause invalid late charges.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Oct 10, 2018, 04:20

@kdjstudios @tim.traylor 

> Thank you for the update. Please look into how SAB is currently processing Net 0. @tim.traylor has been in meetings all day and I have not been able to discuss this with him. I think we will need to have this discussed further in-depth tomorrow morning.

What you deducted Kyle is exactly what code is doing ... terms `0` means invoice to invoice. In other words, the balance should be paid by the time next invoice is generated irrespective of the cycle.  

updated file (only answernet) with 3 additional columns `last_payment_amount`, `last_payment_date`, `discuss`

discuss column will have two entries 'should be charged' and 'blank'

**'blank' means that it needs a manual attention** 

````ruby
if last_payment_date > due_date_per_code
  discuss = 'should be charged'
else
  discuss = "blank"
end
````

[account_with_late_feesanswernet34.xlsx](/uploads/b339a858df4aabe4949c78e20e23ff6d/account_with_late_feesanswernet34.xlsx)

**Comment by kdjstudios** [comment]
In GitLab by @rchauhan on Oct 10, 2018, 07:31

[account_with_late_feesanswernet34-rc.xlsx](/uploads/d66c3fde6a1472e28f0c1040d0cac629/account_with_late_feesanswernet34-rc.xlsx)

**Comment by kdjstudios** [comment]
In GitLab by @rchauhan on Oct 10, 2018, 08:05

[account_with_late_feesanswernet34-rc2.xlsx](/uploads/5efba74577e0e08a411c7c37e7b48f9e/account_with_late_feesanswernet34-rc2.xlsx)

To discuss.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 10, 2018, 10:52

Hello Team,

Just posting an update the recap this morning meeting and discussion. In summary we have found a long time late fee bug; we have also found issue with previous late fee invalid credit list; and we discussed how Net 0 and Manual invoices should work. 

**ACTION PLAN**

1.The Aavaz Team discovered a Long time bug in the calculation for Late Fees.
- I have opened a new ticket: #1191

2.Progress of resolving the Late Fee Bug introduced in #999 
- [Complete] Initial lists generated and sent to sites/clients
- [Complete] Feedback and approvals received back
 - [Complete] Were able to determine that the lists sent out were not accurate.
- [Complete] New lists generated in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1160#note_28412
 - [Complete] I am spot checking and confirming the visual validation accounts.
- [In Progress] Confirmed accurate list approved
 - [Pending] Remove invalid credits for sites that have not processed a billing cycle
 - [Pending] Discuss what to do with sites that have processed a billing cycle.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 10, 2018, 14:25

After reviewing the visual aid checks and spot checking a few accounts. It would appear that all accounts with net terms <= 30 are good to be updated. While the accounts that have <30 need to be handled individually.

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Oct 11, 2018, 06:01

@kdjstudios Hi Kyle, all of the credits that can be reversed have been.  We have two sites, Chicago and Towne that already processed the billing with the incorrect credits:

* Chicago: 18 accounts for $448.00
* Towne: 72 accounts for $390.00

For these two we need to explain to the Operational people that the original assessments were actually valid, and that we can, if they want, generate debits that offset hose credits.  Given the relatively small amount of money they may simply want to let it go and not cause questions on the accounts.

Laser has been re-checked and there are only 9 accounts that should have credits created for them.  I understand we did not yet generate anything for them because they did not provide us the credit code to use. We need to give them the updated list of accounts and request their approval so Aavaz can generate the transactions. Sumeet will be posting the spreadsheet for Laser on the Laser ticket (#1163).  

On Speed-EZ, it turns out that all of the late charges were valid, so there are no adjustments to make for them.  We do want to communicate to them that we backed off the credits and their info is already accurate.

For Keener, they manually addressed their own accounts but they did ask us do double check them.  Sumeet still needs to create an updated/accurate list for them, which he will post on their ticket (#1165).

For Answernet, once you post the updated spreadsheet with the results of the visual check the Aavaz team will have what they need to be able to generate the accurate credits.

Thx

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Oct 11, 2018, 06:06

@smasih A reminder to post the testing results of the temporary code changes (noted at the start of the ticket) for our review.  We need to approve that by Friday so it can go live before more billing cycles occur.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Oct 11, 2018, 06:13

[account_with_late_feeshosted_new_master.csv](/uploads/7bbe0430143856bcbcc84321e199e973/account_with_late_feeshosted_new_master.csv)

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Oct 11, 2018, 06:15

[hosted_late_fees_master.xlsx](/uploads/fdfb1858b22c16a24530664ba4a0691b/hosted_late_fees_master.xlsx)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 11, 2018, 08:10

@smasih @tim.traylor @pchaudhary @amishra 

Oh my apologizes team. I could have sworn I attached my updated research and spot check list. As you can see all accounts that I spot checked with net terms that were equal to or under 30 matched what the Aavaz team found. However the accounts that were marked for visual check (All accounts that had net terms over 30) were each unique in how they need to be handled. 

[account_with_late_feesanswernet34_-__KDJ_10-10-18.xlsx](/uploads/2965004c05bcc544e8542b6bb02d19e7/account_with_late_feesanswernet34_-__KDJ_10-10-18.xlsx)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 11, 2018, 08:31

@tim.traylor @pchaudhary @smasih @amishra 

**Laser:**

> Laser has been re-checked and there are only 9 accounts that should have credits created for them.  I understand we did not yet generate anything for them because they did not provide us the credit code to use. We need to give them the updated list of accounts and request their approval so Aavaz can generate the transactions. Sumeet will be posting the spreadsheet for Laser on the Laser ticket (#1163).

Laser did provide us with a code http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1163#note_28358 and they did provide us with a manual list of valid vs invalid late charges http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1163#note_28284. However it would appear that the list they provided us only shows 17 accounts, while the list Aavaz provided us shows 19 accounts, and @tim.traylor is stating there are only 9 accounts? Something does not seem to be correlating correctly here. I also saw this email this morning, but the post appears to have been deleted from gitlab.[Re_SABilling__Laser_-_SA_Billing_-_Late_Fee_Account_List___1163_.msg](/uploads/eac5e3545fd6120ea4562fe394aedc2b/Re_SABilling__Laser_-_SA_Billing_-_Late_Fee_Account_List___1163_.msg) 

May we please get an update as to what has been done, what needs to be done, and if anything needs to be undone?

**Keener:**

From reading the posts in #1165; It appears that we received a manual list of approved charges on Oct 4th, we then had the Aavaz Team upload two files that needed to be reviewed on Oct 9th, and since we have not follow up with the client they have resent the manual list again on Oct 10th. But from @tim.traylor statement above:

> For Keener, they manually addressed their own accounts but they did ask us do double check them.  Sumeet still needs to create an updated/accurate list for them, which he will post on their ticket (#1165).

It would appear that we are now waiting for a new list to send to the client?

**Speed-Ez:**

I would just like to confirm that we need to inform Speed-Ez that their list of 84 accounts were actually all valid and that we miscalculated all of the accounts on the first list? This does not seem correct, as speed-Ez, already had pending staged fees on 22 accounts, and had already finalized credits on 2 more accounts? Juding from Aavaz's post http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1164#note_28205 

May we please get more of an explanation so I may accurately covey what needs to happen or what needs to not happen to the client.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Oct 11, 2018, 08:51

@kdjstudios 

updated hosted sheet

[hosted_late_fees_master.xlsx](/uploads/e68bb93e963766b897d5f1c8620bdce0/hosted_late_fees_master.xlsx)

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Oct 12, 2018, 05:44

Test cases and results for late fee fix.
http://gitlab.aavaz.biz/AnswerNet/SABilling/wikis/Late-Fees-Testing-Step

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Oct 15, 2018, 05:37

@amishra Hi Asish, the comparison of the AR balance and bad late charge should be <= (less than or equal to).  So if AR balance <= bad late charge, skip the late charging.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 16, 2018, 09:23

If any billing cycles have ran 10/02/18 to today (10/16/18); If so, we will need to push the full late fee fix immediately and recalculate any accounts that have the #999 bug late fee charge.
  - After checking. This can be release tomorrow morning.

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Oct 17, 2018, 06:01

@smasih @kdjstudios  Sumeet is going to check and see if we had any billing cycles run between when the credits were posted and when the code patch went in, as there could be accounts that received an additional late charge when they should not have.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Oct 17, 2018, 07:41

Hi [Tim] ( @tim.traylor ) and [Kyle] ( @kdjstudios ),  

We have successfully released the patch code on both of the production servers and tested as well.

Here is the list of temporary collection accounts.

[final_sheet.csv](/uploads/c2f3639c240c8995f55c41da2b53feff/final_sheet.csv)

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Oct 17, 2018, 07:41

added 5h of time spent at 2018-10-17

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 13, 2018, 14:04

This should now be resolved. Marking for close

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 13, 2018, 14:36

closed
